### PR TITLE
sql,coord: beginnings of ALTER CLUSTER

### PR DESCRIPTION
**Pull request by benesch**
ALTER CLUSTER is a possibly short-lived command to add and remove
replicas of clusters. Specifically, the syntax of CREATE CLUSTER ...
REMOTE is changed to

    CREATE CLUSTER <name> [REMOTE <name> (<host>, <host> ...) ...]

where multiple REMOTE clauses may now be specified to name multiple
replicas. A nearly identical ALTER CLUSTER command

    ALTER CLUSTER <name> [REMOTE <name> (<host>, <host> ...) ...]

can we used to wholesale replace the remote replica set of a cluster.
Replicas are identified by the user provided name. After alteration, any
new replicas are added and any unnamed replicas are removed. Changing
the host specification of an existing replica is not permitted.

Forthcoming work will make the concept of a replica more explicitly part
of the system (e.g., by allowing virtual and managed replicas). But for
now this unblocks the active replication work.

### Motivation

  * This PR adds a known-desirable feature.

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered. (Unstable feature; tests later.)

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - N/A

### CODE REVIEW

File: src/coord/src/coord.rs [hunk 1]
```diff
@@ -2021,6 +2025,18 @@ impl Coordinator {
...
     }
 
+    async fn sequence_alter_compute_instance(
+        &mut self,
+        plan: AlterComputeInstancePlan,
+    ) -> Result<ExecuteResponse, CoordError> {
+        // TODO(benesch): update catalog with new configuration.
+
+        // TODO(mcsherry?): frob the controller with the diff of the compute
+        // instance configuration (`plan.config`).
```
Comment (frankmcsherry) [inline src/coord/src/coord.rs:2035]: Do we need to do anything here? We could just ship the new instance plan at the controller and have it sort things out (i.e. do the diff there, as it is the source of truth of what is installed).

File: src/dataflow-types/src/client.rs [hunk 2]
```diff
@@ -71,8 +71,7 @@ pub enum InstanceConfig {
     },
     /// Out-of-process named instance
     Remote {
-        /// The hosts to connect to.
-        hosts: Vec<String>,
+        replicas: Vec<RemoteReplica>,
         /// Logging configuration.
         logging: Option<LoggingConfig>,
```
Comment (frankmcsherry) [inline src/dataflow-types/src/client.rs:76]: One thing this raises is that the logging stuff isn't going to be well replicated. Peeks and stuff might be ok, but TAIL against any of the introspection sources will be undefined.

**Review by frankmcsherry — COMMENTED** [review]
Looks like a good start. I bet we could just feed the intended config at the compute instance controller and ask it to do the differencing, if we trust that any names that stay put mean the replica is unchanged too (probably not a legit assumption, but we could get started with that).

**Comment by benesch** [comment]
Sweet! I'm going to push up a fix for the failed tests.

**Comment by benesch** [comment]
Great! I fixed the test failures locally and marked for automerge. No substantive changes, so I squashed into a single commit with a proper commit message.

`test/cluster` has a working example of using `ALTER CLUSTER`. Well, the test is still flaky for reasons unrelated to `ALTER CLUSTER`, but if you manage to avoid that flake, the addition of the second replica seems to work!

**Pull request merged**